src/include/baselib/jni/JavaVirtualMachine.h

```diff
@@ -0,0 +1,178 @@
+                            << jniReturn
+                        );
+                }
+            }
+
+        private:
+
+            static cpp::ScalarTypeIniter< bool >                        g_javaVMCreated;
+            static os::mutex                                            g_lock;
+
+            const bl::os::library_ref                                   m_jvmLibrary;
+            JavaVM*                                                     m_javaVM;
+
+            std::string getJvmPathFromJavaHome()
+            {
+                const auto javaHome = bl::os::tryGetEnvironmentVariable( "JAVA_HOME" );
+
+                BL_CHK_USER_FRIENDLY(
+                    true,
+                    ! javaHome.get(),
+                    "Environment variable JAVA_HOME is not defined"
+                    );
+
+                bl::fs::path jvmPath( *javaHome );
+
+                if( bl::os::onWindows() )
+                {
+                    jvmPath += "/jre/bin/server/jvm.dll";
+                }
+                else
+                {
+                    jvmPath += "/jre/lib/amd64/server/libjvm.so";
+                }
+
+                jvmPath = bl::fs::normalize( jvmPath );
+
+                BL_CHK_USER_FRIENDLY(
+                    false,
+                    bl::fs::exists( jvmPath ),
+                    BL_MSG()
+                        << "Path doesn't exist "
+                        << bl::fs::normalizePathParameterForPrint( jvmPath )
+                    );
+
+                if( bl::os::onWindows() )
+                {
+                    /*
+                     * Remove "\\?\" prefix from JavaVM path on Windows.
+                     * Otherwise after the JVM library is loaded the call to JNI_CreateJavaVM will fail with:
+                     *     Error occurred during initialization of VM
+                     *     java/lang/NoClassDefFoundError: java/lang/Object
+                     */
+
+                    return bl::fs::detail::WinLfnUtils::chk2RemovePrefix( std::move( jvmPath ) ).string();
+                }
+                else
+                {
+                    return jvmPath.string();
+                }
+            }
+
+            void createJvm()
+            {
+                BL_MUTEX_GUARD( g_lock );
+
+                /*
+                 * Creation of multiple VMs in a single process is not supported.
+                 * http://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/invocation.html#JNI_CreateJavaVM
+                 */
+
+                BL_CHK_USER_FRIENDLY(
+                    true,
+                    g_javaVMCreated,
+                    "JavaVM has already been created"
+                    );
+
+                const auto procAddress = bl::os::getProcAddress( m_jvmLibrary, "JNI_CreateJavaVM" );
+
+                const auto jniCreateJavaVM = reinterpret_cast< jint ( JNICALL* )( JavaVM**, void**, void *) >( procAddress );
+
+                JavaVMInitArgs vmArgs;
```

Comment:
All fields of JavaVMInitArgs are set here (VC14 even gives a warning if any field remains unset). I'll add a default struct initializer "= {};" to make it clear and in case new fields are added in the future.